src/Microsoft.Framework.Caching.Memory/MemoryCache.cs

```diff
@@ -62,27 +62,51 @@ public int Count
             get { return _entries.Count; }
         }
 
-        public object Set([NotNull] string key, IEntryLink link, object state, [NotNull] Func<ICacheSetContext, object> create)
+        public object Set([NotNull] string key, object value, IEntryLink link, CacheEntryOptions cacheEntryOptions)
         {
             CheckDisposed();
             CacheEntry priorEntry = null;
-            var now = _clock.UtcNow;
-            var context = new CacheSetContext(key) { State = state, CreationTime = now };
-            object value = create(context);
-            var entry = new CacheEntry(context, value, _entryExpirationNotification);
+            var utcNow = _clock.UtcNow;
+
+            DateTimeOffset? absoluteExpiration = null;
+            if (cacheEntryOptions.AbsoluteExpirationRelativeToNow.HasValue)
+            {
+                absoluteExpiration = utcNow + cacheEntryOptions.AbsoluteExpirationRelativeToNow;
+            }
+            else if (cacheEntryOptions.AbsoluteExpiration.HasValue)
+            {
+                if (cacheEntryOptions.AbsoluteExpiration <= utcNow)
+                {
+                    throw new ArgumentOutOfRangeException(
```

Comment:
The problem is with the value of `utc now` which is calculated when the cache entry is set...also if we move the `utc now` to cache entry options, its problematic as there could be time delay between when the options were created vs when the entry was set and also the cache entry options could be cached and used for different entries...